internal/controller/vrg_volrep.go

```diff
@@ -1289,9 +1291,28 @@ func (v *VRGInstance) selectVolumeReplicationClass(
 		}
 
 		// ReplicationClass that matches both VRG schedule and pvc provisioner
-		if schedulingInterval == v.instance.Spec.Async.SchedulingInterval {
-			matchingReplicationClassList = append(matchingReplicationClassList, replicationClass)
+		if schedulingInterval != v.instance.Spec.Async.SchedulingInterval {
+			continue
 		}
+
+		// if peerClasses does not exist, replicationClasses would not have SID in
+		// older ramen versions, this check is neeed because we need to handle upgrade
+		// scenario where SID is not present in replicatioClass.
+
+		// if peerClass exist, continue to check if SID matches, or skip the check and proceed
+		// to append to matchingReplicationClassList
+		if len(v.instance.Spec.Async.PeerClasses) != 0 {
+			sIDFromReplicationClass, exists := replicationClass.GetLabels()[StorageIDLabel]
+			if !exists {
+				continue
+			}
+
+			if sIDFromReplicationClass != storageClass.GetLabels()[StorageIDLabel] {
+				continue
+			}
+		}
+
+		matchingReplicationClassList = append(matchingReplicationClassList, replicationClass)
```

Comment:
`break` right after, given that we only need one match.

Reply:
@BenamarMk this is present such that if there are more than on VRClass pointing to the same SC, we pick either using the VRClass label selector or pick the "default" VRClass. This was added to deal with a VRClass for VMs that includes flatten option, and so we needed a default label to proceed.

Unless I am missing some other case that you noticed.